<?xml version="1.0"?>
<!-- For visualization in real experiments -->

<launch>
    <arg name="namespace" default="uav1"/>
    <arg name="open_rviz" default="true"/>

    <arg name="visualize_max_freq" default="100"/> <!-- (Hz) Maximum  Maximum base_link tf publishing rate  -->
    <arg name="enable_history_path" default="true"/> <!-- Enable visualizing history path  -->
    <arg name="visualize_path_time" default="30"/> <!-- (s) displaying history path time -->
    <arg name="visualize_tf_frame" default="map"/> <!-- tf world frame name -->
    <arg name="visualize_marker_name" default="$(arg namespace)"/> <!-- visualize_marker_name -->
    <arg name="base_link_name" default="base_link"/> <!-- base link name in the robot urdf -->
    <arg name="base_link_tf_prefix" default="$(arg namespace)"/> <!-- tf prefix added to the base_link_name -->
    <arg name="rotor_0_joint_name" default="rotor_0_joint"/> <!-- rotor revolute joint name in the robot urdf -->
    <arg name="rotor_1_joint_name" default="rotor_1_joint"/> <!-- rotor revolute joint name in the robot urdf -->
    <arg name="rotor_2_joint_name" default="rotor_2_joint"/> <!-- rotor revolute joint name in the robot urdf -->
    <arg name="rotor_3_joint_name" default="rotor_3_joint"/> <!-- rotor revolute joint name in the robot urdf -->
    <arg name="local_pos_source" default="0"/> <!-- mavros/local_position/pose source -->     
    <!-- 0 for mocap (use local_position/pose for visualization directly) -->     
    <!-- 1 for GPS (use global_position/global relative to the world_origin_gps for x,y visualization, use local_position/pose for z and attitude visualization) -->
    <!-- world origin gps values as the ref of visualizing the local position if local_pos_source = 1 -->
    <arg name="world_origin_latitude_deg" default="39.978861"/> <!-- degree -->
    <arg name="world_origin_longitude_deg" default="116.339803"/> <!-- degree -->
    <arg name="world_origin_AMSL_alt_metre" default="53.0"/> <!-- metre (Above Mean Sea Level) (not WGS84/ELlipsoid of GPS) -->
    <arg name="urdf_model_path" default="$(find px4_rotor_sim)/model/iris.urdf"/>

<!-- For visualization in real experiments -->
<group ns="$(arg namespace)">    
    <param name="robot_description" textfile="$(arg urdf_model_path)" />
    
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" >
        <param name="tf_prefix" value="$(arg namespace)"/>
    </node>
    
    <node pkg="px4_rotor_sim" type="px4_rotor_visualizer_node" name="px4_rotor_visualizer_node$(arg namespace)" output="screen">
        <param name="visualize_max_freq" value="$(arg visualize_max_freq)"/>
        <param name="visualize_path_time" value="$(arg visualize_path_time)"/>
        <param name="visualize_tf_frame" value="$(arg visualize_tf_frame)"/>
        <param name="visualize_marker_name" value="$(arg visualize_marker_name)"/>
        <param name="base_link_name" value="$(arg base_link_name)"/>
        <param name="base_link_tf_prefix" value="$(arg base_link_tf_prefix)"/>
        <param name="rotor_0_joint_name" value="$(arg rotor_0_joint_name)"/>
        <param name="rotor_1_joint_name" value="$(arg rotor_1_joint_name)"/>
        <param name="rotor_2_joint_name" value="$(arg rotor_2_joint_name)"/>
        <param name="rotor_3_joint_name" value="$(arg rotor_3_joint_name)"/>
        <param name="local_pos_source" value="$(arg local_pos_source)"/>
        <param name="world_origin_latitude_deg" value="$(arg world_origin_latitude_deg)"/>
        <param name="world_origin_longitude_deg" value="$(arg world_origin_longitude_deg)"/>
        <param name="world_origin_AMSL_alt_metre" value="$(arg world_origin_AMSL_alt_metre)"/>
    </node>
    
    <group if="$(arg open_rviz)">
        <node name="rviz" pkg="rviz" type="rviz" args="-d $(find px4_rotor_sim)/launch/single_px4_rotor.rviz" required="true" />
    </group>
</group>

</launch>
